File: sdks/typescript/src/quantum-local.ts
export interface QuantumLocalAccelerator {
  backend: "customer_gpu" | string;
  run(
    probabilities: readonly number[],
    shots: number,
    seed: number,
  ): Promise<Record<string, number>>;
}

export interface QuantumLocalResult {
  selected_index: number;
  selected: string;
  probabilities: number[];
  /** Real amplitudes for the prepared, normalized state; padded to 2^qubits. */
  amplitudes: number[];
  qubits: number;
  counts: Record<string, number>;
  shots: number;
  backend: string;
}

function validateProviderCounts(counts: Record<string, number>, shots: number): void {
  const entries = Object.entries(counts);
  if (
    entries.length === 0 ||
    entries.some(
      ([key, value]) => !/^[01]{1,64}$/.test(key) || !Number.isInteger(value) || value < 0,
    ) ||
    entries.reduce((sum, [, value]) => sum + value, 0) !== shots
  ) {
    throw new Error("provider returned invalid counts");
  }
}

function finite(value: number, name: string): number {
  if (!Number.isFinite(value)) throw new TypeError(`${name} must be finite`);
  return value;
}

function validateCounts(counts: Record<string, number>, size: number, shots: number): void {
  const entries = Object.entries(counts);
  if (
    entries.length === 0 ||
    entries.some(
      ([key, value]) =>
        !/^\d+$/.test(key) || Number(key) >= size || !Number.isInteger(value) || value < 0,
    ) ||
    entries.reduce((sum, [, value]) => sum + value, 0) !== shots
  ) {
    throw new Error("accelerator returned invalid counts");
  }
}

export async function quantumLocalSelect(input: {
  candidates: readonly string[];
  scores: readonly number[];
  shots?: number;
  seed?: number;
  accelerator?: QuantumLocalAccelerator;
}): Promise<QuantumLocalResult> {
  const { candidates, scores } = input;
  if (candidates.length < 1 || candidates.length > 256 || candidates.length !== scores.length) {
    throw new RangeError("candidates and scores must have the same length from 1 to 256");
  }
  if (candidates.some((value) => !value || value.length > 256_000))
    throw new RangeError("invalid candidate");
  for (const value of scores) finite(value, "score");
  const shots = input.shots ?? 2_000;
  const seed = input.seed ?? 0xbee;
  if (!Number.isInteger(shots) || shots < 1 || shots > 100_000)
    throw new RangeError("shots out of range");
  if (!Number.isSafeInteger(seed)) throw new RangeError("seed must be a safe integer");
  const high = Math.max(...scores);
  const weights = scores.map((score) => Math.exp(score - high));
  const total = weights.reduce((sum, value) => sum + value, 0);
  const probabilities = weights.map((value) => value / total);
  const qubits = Math.max(1, Math.ceil(Math.log2(candidates.length)));
  const amplitudes = Array.from({ length: 2 ** qubits }, (_, index) =>
    index < probabilities.length ? Math.sqrt(probabilities[index]) : 0,
  );
  let counts: Record<string, number>;
  let backend = "customer_cpu";
  if (input.accelerator) {
    counts = await input.accelerator.run(probabilities, shots, seed);
    backend = input.accelerator.backend;
  } else {
    counts = {};
    let state = seed >>> 0;
    for (let shot = 0; shot < shots; shot += 1) {
      state = (Math.imul(1664525, state) + 1013904223) >>> 0;
      const sample = state / 2 ** 32;
      let cumulative = 0;
      let index = probabilities.length - 1;
      for (let candidate = 0; candidate < probabilities.length; candidate += 1) {
        cumulative += probabilities[candidate];
        if (sample < cumulative) {
          index = candidate;
          break;
        }
      }
      counts[String(index)] = (counts[String(index)] ?? 0) + 1;
    }
  }
  validateCounts(counts, candidates.length, shots);
  const selectedIndex = Number(
    Object.keys(counts).reduce((best, key) => (counts[key] > counts[best] ? key : best)),
  );
  return {
    selected_index: selectedIndex,
    selected: candidates[selectedIndex],
    probabilities,
    amplitudes,
    qubits,
    counts,
    shots,
    backend,
  };
}

export interface ByopaDirectRequest {
  circuit: string;
  shots: number;
  backend: string;
  max_cost_minor: number;
  currency: string;
}
export interface ByopaDirectResult {
  provider_job_id: string;
  backend: string;
  counts: Record<string, number>;
  billed_cost_minor?: number;
}
export interface ByopaDirectAdapter {
  quote(
    request: ByopaDirectRequest,
  ): Promise<{ cost_minor: number; currency: string; expires_at: string }>;
  run(
    request: ByopaDirectRequest,
    quote: { cost_minor: number; currency: string; expires_at: string },
  ): Promise<ByopaDirectResult>;
}

export async function executeByopaDirect(
  request: ByopaDirectRequest,
  adapter: ByopaDirectAdapter,
): Promise<ByopaDirectResult> {
  if (!request.circuit || request.circuit.length > 1_000_000)
    throw new RangeError("invalid circuit");
  if (!Number.isInteger(request.shots) || request.shots < 1 || request.shots > 100_000)
    throw new RangeError("invalid shots");
  if (!request.backend || request.backend.length > 200) throw new RangeError("invalid backend");
  if (!Number.isSafeInteger(request.max_cost_minor) || request.max_cost_minor < 0)
    throw new RangeError("invalid cost cap");
  if (!/^[A-Z]{3}$/.test(request.currency))
    throw new RangeError("currency must be uppercase ISO-4217");
  const quote = await adapter.quote(request);
  if (
    !Number.isSafeInteger(quote.cost_minor) ||
    quote.cost_minor < 0 ||
    quote.cost_minor > request.max_cost_minor ||
    quote.currency !== request.currency ||
    !/(?:Z|[+-]\d{2}:\d{2})$/.test(quote.expires_at) ||
    !Number.isFinite(Date.parse(quote.expires_at)) ||
    Date.parse(quote.expires_at) <= Date.now()
  )
    throw new Error("provider returned invalid quote");
  const result = await adapter.run(request, quote);
  if (
    !result.provider_job_id ||
    result.provider_job_id.length > 300 ||
    result.backend !== request.backend
  )
    throw new Error("provider returned invalid identity");
  validateProviderCounts(result.counts, request.shots);
  if (
    result.billed_cost_minor !== undefined &&
    (!Number.isSafeInteger(result.billed_cost_minor) ||
      result.billed_cost_minor < 0 ||
      result.billed_cost_minor > quote.cost_minor ||
      result.billed_cost_minor > request.max_cost_minor)
  )
    throw new Error("provider returned invalid bill");
  return result;
}

File: sdks/python/bee_sdk/quantum_local.py
"""Customer-local statevector selection and BYOPA-direct validation (stdlib only)."""

from __future__ import annotations

import math
import random
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Protocol


@dataclass(frozen=True)
class QuantumLocalResult:
    selected_index: int
    selected: str
    probabilities: list[float]
    amplitudes: list[float]
    qubits: int
    counts: dict[str, int]
    shots: int
    backend: str


def _valid_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def quantum_local_select(
    candidates: list[str], scores: list[float], shots: int = 2_000, seed: int = 0xBEE
) -> QuantumLocalResult:
    if not 1 <= len(candidates) <= 256 or len(scores) != len(candidates):
        raise ValueError("candidates and scores must have the same length from 1 to 256")
    if any(not value or len(value) > 256_000 for value in candidates):
        raise ValueError("invalid candidate")
    if any(isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) for value in scores):
        raise ValueError("scores must be finite numbers")
    if not _valid_int(shots) or not 1 <= shots <= 100_000 or not _valid_int(seed):
        raise ValueError("shots or seed out of range")
    high = max(scores)
    weights = [math.exp(score - high) for score in scores]
    total = sum(weights)
    probabilities = [value / total for value in weights]
    qubits = max(1, math.ceil(math.log2(len(candidates))))
    amplitudes = [
        math.sqrt(probabilities[index]) if index < len(probabilities) else 0.0
        for index in range(2**qubits)
    ]
    rng = random.Random(seed)
    counts: dict[str, int] = {}
    for _ in range(shots):
        sample = rng.random()
        cumulative = 0.0
        index = len(probabilities) - 1
        for candidate, probability in enumerate(probabilities):
            cumulative += probability
            if sample < cumulative:
                index = candidate
                break
        counts[str(index)] = counts.get(str(index), 0) + 1
    selected_index = int(max(counts, key=counts.get))  # type: ignore[arg-type]
    return QuantumLocalResult(
        selected_index,
        candidates[selected_index],
        probabilities,
        amplitudes,
        qubits,
        counts,
        shots,
        "customer_cpu",
    )


class ByopaDirectAdapter(Protocol):
    def quote(self, request: dict) -> dict: ...
    def run(self, request: dict, quote: dict) -> dict: ...


def execute_byopa_direct(request: dict, adapter: ByopaDirectAdapter) -> dict:
    circuit, shots = request.get("circuit"), request.get("shots")
    backend, cap, currency = request.get("backend"), request.get("max_cost_minor"), request.get("currency")
    if not isinstance(circuit, str) or not circuit or len(circuit) > 1_000_000:
        raise ValueError("invalid circuit")
    if not _valid_int(shots) or not 1 <= shots <= 100_000:
        raise ValueError("invalid shots")
    if not isinstance(backend, str) or not backend or len(backend) > 200:
        raise ValueError("invalid backend")
    if not _valid_int(cap) or cap < 0 or not isinstance(currency, str) or len(currency) != 3 or not currency.isupper():
        raise ValueError("invalid cost cap or currency")
    quote = adapter.quote(request)
    expires = quote.get("expires_at")
    try:
        expiry = datetime.fromisoformat(str(expires).replace("Z", "+00:00"))
    except ValueError as error:
        raise ValueError("provider returned invalid quote") from error
    if expiry.tzinfo is None:
        raise ValueError("provider returned invalid quote")
    cost = quote.get("cost_minor")
    if not _valid_int(cost) or cost < 0 or cost > cap or quote.get("currency") != currency or expiry <= datetime.now(timezone.utc):
        raise ValueError("provider returned invalid quote")
    result = adapter.run(request, quote)
    counts = result.get("counts")
    if (
        not isinstance(result.get("provider_job_id"), str)
        or not result["provider_job_id"]
        or result.get("backend") != backend
        or not isinstance(counts, dict)
        or not counts
        or any(
            not isinstance(key, str)
            or not 1 <= len(key) <= 64
            or set(key) - {"0", "1"}
            or not _valid_int(value)
            or value < 0
            for key, value in counts.items()
        )
        or sum(counts.values()) != shots
    ):
        raise ValueError("provider returned invalid result")
    billed = result.get("billed_cost_minor")
    if billed is not None and (not _valid_int(billed) or billed < 0 or billed > cost or billed > cap):
        raise ValueError("provider returned invalid bill")
    return result
